### PR TITLE
raftstore: encrypt lock cf in snapshots

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?

Complete encryption for snapshots.

Based on #7534.

### What is changed and how it works?

When generating cf file for CF lock, encrypt the content. And when applying it, decrypt the content.

### Check List 

Tests 

- Unit test

Side effects

### Release note

### CODE REVIEW

File: components/raftstore/src/store/snap/io.rs [hunk 1]
```diff
@@ -163,6 +207,27 @@ where
...
 
+pub fn get_decrypter_reader(
+    file: &str,
+    encryption_key_manager: &DataKeyManager,
+) -> Result<Box<dyn Read + Send>, Error> {
+    let enc_info = box_try!(encryption_key_manager.get_file(file));
+    let mthd = encryption_method_from_db_encryption_method(enc_info.method);
+    debug!(
+        "get_decrypter_reader gets enc_infor for {:?}, method: {:?}", file, mthd;
+        "key" => hex::encode_upper(&enc_info.key),
```
Comment (overvenus) [inline components/raftstore/src/store/snap/io.rs:218]: Please do not log **key**.

File: components/encryption/src/io.rs [hunk 2]
```diff
@@ -203,23 +219,19 @@ impl<W: Write> Write for EncrypterWriter<W> {
...
 impl<W: Write> Drop for EncrypterWriter<W> {
     fn drop(&mut self) {
-        let mut encrypt_buffer = vec![0; self.block_size];
-        let bytes = self.crypter.finalize(&mut encrypt_buffer).unwrap();
-        if bytes != 0 {
-            // The EncrypterWriter current only support crypters that always return the same amount
-            // of data. This is true for CTR mode.
-            panic!("unsupported encryption");
-        }
+        self.do_finalize();
```
Comment (yiwu-arbug) [inline components/encryption/src/io.rs:235]: it looks like if `finalize()` is called, `do_finalize()` (as well as `crypter.finalize()`) will be called twice (the second time in `drop()`). can we avoid calling `crypter.finalize()` a second time?

File: components/raftstore/src/store/snap/io.rs [hunk 3]
```diff
@@ -36,15 +43,45 @@ pub fn build_plain_cf_file<E>(
...
         stats.key_count += 1;
         stats.total_size += key.len() + value.len();
-        box_try!(file.encode_compact_bytes(key));
-        box_try!(file.encode_compact_bytes(value));
+        box_try!(BytesEncoder::encode_compact_bytes(&mut writer, key));
+        box_try!(BytesEncoder::encode_compact_bytes(&mut writer, value));
         Ok(true)
     }));
+
+    let mut file = match file {
```
Comment (yiwu-arbug) [inline components/raftstore/src/store/snap/io.rs:80]: the code would be more readable if we add a `should_encrypt` boolean variable earlier in this function.

File: components/encryption/src/io.rs (hunk 2 above)
Comment (hicqu) [reply components/encryption/src/io.rs:235]: God catch.

File: components/encryption/src/io.rs [hunk 4]
```diff
@@ -186,20 +186,23 @@ impl<W: Write> EncrypterWriter<W> {
...
-        if bytes != 0 {
-            // The EncrypterWriter current only support crypters that always return the same amount
-            // of data. This is true for CTR mode.
-            panic!("unsupported encryption");
+        self.do_finalize().unwrap()
+    }
+
+    fn do_finalize(&mut self) -> Option<W> {
+        if self.writer.is_some() {
+            drop(self.flush());
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/encryption/src/io.rs:196]: `flush` returns a `Result`, which needs to be comsumed. Otherwise clippy will warn it.

**Pull request merged**